src/System.Security.Cryptography.Cng/src/System/Security/Cryptography/CngAlgorithm.cs

```diff
@@ -0,0 +1,290 @@
+
+                if (s_ecdsap256 == null)
+                {
+                    s_ecdsap256 = new CngAlgorithm("ECDSA_P256"); // BCRYPT_ECDSA_P256_ALGORITHM
+                }
+
+                return s_ecdsap256;
+            }
+        }
+
+        public static CngAlgorithm ECDsaP384
+        {
+            get
+            {
+                Contract.Ensures(Contract.Result<CngAlgorithm>() != null);
+
+                if (s_ecdsap384 == null)
+                {
+                    s_ecdsap384 = new CngAlgorithm("ECDSA_P384"); // BCRYPT_ECDSA_P384_ALGORITHM
+                }
+
+                return s_ecdsap384;
+            }
+        }
+
+        public static CngAlgorithm ECDsaP521
+        {
+            get
+            {
+                Contract.Ensures(Contract.Result<CngAlgorithm>() != null);
+
+                if (s_ecdsap521 == null)
+                {
+                    s_ecdsap521 = new CngAlgorithm("ECDSA_P521"); // BCRYPT_ECDSA_P521_ALGORITHM
+                }
+
+                return s_ecdsap521;
+            }
+        }
+
+        public static CngAlgorithm MD5
+        {
+            get
+            {
+                Contract.Ensures(Contract.Result<CngAlgorithm>() != null);
+
+                if (s_md5 == null)
+                {
+                    s_md5 = new CngAlgorithm("MD5"); // BCRYPT_MD5_ALGORITHM
+                }
+
+                return s_md5;
+            }
+        }
+
+        public static CngAlgorithm Sha1
+        {
+            get
+            {
+                Contract.Ensures(Contract.Result<CngAlgorithm>() != null);
+
+                if (s_sha1 == null)
+                {
+                    s_sha1 = new CngAlgorithm("SHA1"); // BCRYPT_SHA1_ALGORITHM
+                }
+
+                return s_sha1;
+            }
+        }
+
+        public static CngAlgorithm Sha256
+        {
+            get
+            {
+                Contract.Ensures(Contract.Result<CngAlgorithm>() != null);
+
+                if (s_sha256 == null)
+                {
+                    s_sha256 = new CngAlgorithm("SHA256"); // BCRYPT_SHA256_ALGORITHM
+                }
+
+                return s_sha256;
+            }
+        }
+
+        public static CngAlgorithm Sha384
+        {
+            get
+            {
+                Contract.Ensures(Contract.Result<CngAlgorithm>() != null);
+
+                if (s_sha384 == null)
+                {
+                    s_sha384 = new CngAlgorithm("SHA384"); // BCRYPT_SHA384_ALGORITHM
+                }
+
+                return s_sha384;
+            }
+        }
+
+        public static CngAlgorithm Sha512
+        {
+            get
+            {
+                Contract.Ensures(Contract.Result<CngAlgorithm>() != null);
+
+                if (s_sha512 == null)
+                {
+                    s_sha512 = new CngAlgorithm("SHA512"); // BCRYPT_SHA512_ALGORITHM
+                }
+
+                return s_sha512;
+            }
+        }
```

Comment:
I think apps expect the static properties to have perf characteristics like using a enum literal. Actually, we could probably replace all this lazy initialization code with static field initializers. 
That would also guarantee a single-instance (though I don't believe single instancing is actually required here.)